### ISSUE TITLE
Embedding arrays into NeuralPDE differential equations

### ISSUE THREAD

**Issue by francescocalisto** [SciML/NeuralPDE.jl#177]
There is currently no way to use an array-derived quantity in a NeuralPDE differential equation.

This is an example:

```
dx = 0.05
dy = 0.05
xwidth   = 10.0    
ywidth   = 10.0
xs = -xwidth*0.5: dx : xwidth*0.5
ys = -ywidth*0.5: dy : ywidth*0.5
xGradMatrix = rand(length(xs),length(ys))  #rand for explainatory purpose
yGradMatrix = rand(length(xs),length(ys))
getDxz(x,y) = xGradMatrix[convert(Int64,x/dx), convert(Int64,y/dy)]
getDyz(x,y) = yGradMatrix[convert(Int64,x/dx), convert(Int64,y/dy)]
S = x^2 * getDxz(x,y) + 3.2 * getDyz(x,y)
eq = Dt(u(t,x,y,θ)) + S*gn ~ 0  
```
This doesn’t work at the moment. What one should be able to do is to get values from xGradMatrix and yGradMatrix as a function of the position.

**Comment by 00krishna** [comment]
@KirillZubov do you know of any updates on this issue? I was just trying to understand what the underlying problem was with including intermediate expressions in one of the system equations?

**Comment by KirillZubov** [comment]
so as we generate data inside, that means that every iteration it new set of data of random, quasi-random number, adaptive quadrature's  adaptively selects the set and the size of the set may be different on a new iteration
https://neuralpde.sciml.ai/dev/solvers/pinns/#Training-strategy
and we can't just input data  in general, it should be respected to  variables - x,y

not work:
```julia
data = rand(2,10)
eq = u(x,y) ~ data
``` 
 it can be work like some interpolation, but probably it bad idea  
```julia
function data(x,y)
     size_ = size(vcat(x,y) )   
     generate_own_data(size_)
end
@register  data(x,y)
eq = u(x,y) ~ data(x,y)
```

so if you have a function with respect to variable f(x,y), so just use it
if it isn't Symbolic, so use @register f(x,y)

probably the best way to insert data is additional_loss. there is the example with parameter estimation
I try to add all this in docs in the next update that will soon

**Comment by KirillZubov** [comment]
probably I think it is a bad idea to inject data in a high level of eq. 
to use the additional loss is bestю And probably the issue can be close
 
mention @ChrisRackauckas

**Comment by ChrisRackauckas** [comment]
Yeah, is the data actually part of the model or the training process? Those are two different things, and I think this issue is just from confusing the two. For this kind of thing, instead it would be better to directly solve the stochastic PDE, which isn't something we can do right now but is better than solving for a specific random number sequence.

**Comment by KirillZubov** [comment]
it will be cool to have an example with ground-truth data, like interpolating some operator in pde from real-world data. 
but the big question is how to connect it together math + data, inside neuralpde some effective way.

**Comment by ChrisRackauckas** [comment]
You just define an appropriate additional loss term.

**Comment by 00krishna** [comment]
I am trying to understand how I can help with this, @ChrisRackauckas @KirillZubov ? In the case of this particular problem, the case is wildfire spread. So there is a 2 dimensional array where each pixel represents the fuel available to the fire. So since this is a level set problem, the boundary will evolve probably in proportion to the curvature at each mesh point and the fuel available at each mesh point. 

So in that sense, the data is not directly coming from the system of equation, it is separate auxiliary data that affects the solution of the system. You can think of it as a component of the initial conditions in some sense. 

For solving the stochastic PDE is that something that we can approach with a MOL discretization and then evolve the time domain with an SDE solver? Or is that like the naive path and no one should do that. 

In looking at the parameter estimation tutorial, it is still not completely clear how to implement it in this case. So that tutorial shows how given a system of equations and some data, how to optimize the parameters by minimizing loss. Now I am not an authority on level set methods and front propagation, so hopefully someone knows this better than I do. But if we have the data and a handful of parameters, is the right way to approach this just estimating the $\theta$ vector values parameter in the original issue post. That just means in the additional loss function, we would have to iterate over the data array for the fuel as well as the difference between the actual front propagation points and the predicted propagation points?

**Comment by ChrisRackauckas** [comment]
If it's not just a loss term, then turn the data into a function using an interpolation and register the interpolation.
